sphinx_needs/need_item.py

```diff
@@ -0,0 +1,333 @@
+# To maintain back-compatibility,
+# we need to implement MutableMapping semantics, these are the methods that need to be implemented:
+# - implement immutable: __getitem__, __iter__, __len__
+# - implement mutable: __setitem__, __delitem__
+# - inherit immutable: __contains__, keys, items, values, get, __eq__, and __ne__
+# - inherit mutable: pop, popitem, clear, update, and setdefault
+#
+# For NeedPartItem, we should not allow any mutability, as it is a view into a need.
+# For NeedItem, we allow mutability, but only for values, i.e. it should not allow adding or removing keys.
+from __future__ import annotations
+
+from collections.abc import Iterable, Iterator, Mapping, Sequence
+from typing import Any, Literal, overload
+
+from sphinx_needs.data import NeedsInfoType, NeedsPartType
+
+
+class NeedItem:
+    """A class representing a single need item."""
+
+    __slots__ = ("_data",)
+
+    _not_required = {"pre_content", "post_content", "constraints_error"}
+    _immutable = {
+        "id",
+        "type",
+        "docname",
+        "lineno",
+        "lineno_content",
+        "id_complete",
+        "id_parent",
+        "is_need",
+        "is_part",
+        "is_external",
+        "is_import",
+    }
+
+    def __init__(self, data: NeedsInfoType) -> None:
+        """Initialize the NeedItem instance."""
+        self._data = data
+
+    def __repr__(self) -> str:
+        """Return a string representation of the NeedItem."""
+        return f"NeedItem({self._data!r})"
+
+    def __str__(self) -> str:
+        """Return a string representation of the NeedItem."""
+        return f"NeedItem({self._data!s})"
+
+    def copy(self) -> NeedItem:
+        """Return a copy of the NeedItem."""
+        return NeedItem(self._data.copy())
+
+    def __eq__(self, other: object) -> bool:
+        """Check if two NeedItems are equal."""
+        if not isinstance(other, NeedItem):
+            return False
+        return self._data == other._data
+
+    def __ne__(self, other: object) -> bool:
+        """Check if two NeedItems are not equal."""
+        return not self.__eq__(other)
+
+    def __contains__(self, key: str) -> bool:
+        """Check if the need item contains a key."""
+        return key in self._data
+
+    def __iter__(self) -> Iterator[str]:
+        """Return an iterator over the keys of the need item."""
+        return self._data.__iter__()
+
+    @overload
+    def __getitem__(
+        self,
+        key: Literal[
+            "collapse",
+            "hide",
+            "is_import",
+            "is_external",
+            "is_modified",
+            "is_need",
+            "is_part",
+            "jinja_content",
+            "has_dead_links",
+            "has_forbidden_dead_links",
+            "constraints_passed",
+        ],
+    ) -> bool: ...
+
+    @overload
+    def __getitem__(
+        self,
+        key: Literal[
+            "id",
+            "id_parent",
```

Comment:
id_parent should be optional, it's about need part parent - would it break the data model to change it to `str | None` ?

Reply:
This particular PR is not intended to change any schema specifics; e.g. it shows that this change does not change any output needs.json
So it may well be a legitimate change, but should not be made here